txpool/pool.go

```diff
@@ -619,41 +625,47 @@ func (p *TxPool) validateTxs(txs TxSlots) (reasons []DiscardReason, goodTxs TxSl
 		return reasons, goodTxs, err
 	}
 
-	j := 0
-	for i := range txs.txs {
-		reasons[i] = p.validateTx(txs.txs[i], txs.isLocal[i])
-		if reasons[i] != Success {
-			if reasons[i] == Spammer {
-				p.punishSpammer(txs.txs[i].senderID)
-			}
+	goodCount := 0
+	for i, txn := range txs.txs {
+		reason := p.validateTx(txn, txs.isLocal[i])
+		if reason == Success {
+			goodCount++
 			continue
 		}
-		reasons[i] = NotSet
-		goodTxs.Resize(uint(j + 1))
-		goodTxs.txs[j] = txs.txs[i]
-		goodTxs.isLocal[j] = txs.isLocal[i]
-		copy(goodTxs.senders.At(j), txs.senders.At(i))
-		j++
+		if reason == Spammer {
+			p.punishSpammer(txn.senderID)
+		}
+		reasons[i] = reason
+	}
+
+	goodTxs.Resize(uint(goodCount))
+
+	j := 0
+	for i, txn := range txs.txs {
+		if reasons[i] == NotSet {
```

Comment:
Should not this be `reasons[i] == Success?` Because now nothing is setting it to `NotSet` above

Reply:
I agree this one is a little unintuitive. `NotSet` is the default value when the `reasons` slice is created, and `validateTxs()` is expected to leave all successes as `NotSet` for the return value, so it only changes a value for a discard.

Of the callers to `validateTxs()`, `OnNewBlock()` and `processRemoteTxs()` both ignore `reasons`, which leaves the other caller, `AddLocalTxs()`.

`AddLocalTxs()` calls `fillDiscardReasons()`, which is the "final" filler -- it looks at only the `NotSet` txs to see if there's any other cause to discard them (`discardReasonsLRU` by `IdHash`), and otherwise sets the final `Success` state.

In other words, a txn is not `Success` until all checks are done, and until then it is inconclusive with `NotSet`.

Reply:
Thanks a lot! Now I see it